response.go

```diff
@@ -103,185 +156,375 @@ func (info *SQLInfo) DecodeMsgpack(d *msgpack.Decoder) error {
 	return nil
 }
 
-func (resp *Response) smallInt(d *msgpack.Decoder) (i int, err error) {
-	b, err := resp.buf.ReadByte()
+func smallInt(d *msgpack.Decoder, buf *smallBuf) (i int, err error) {
+	b, err := buf.ReadByte()
 	if err != nil {
 		return
 	}
 	if b <= 127 {
 		return int(b), nil
 	}
-	resp.buf.UnreadByte()
+	buf.UnreadByte()
 	return d.DecodeInt()
 }
 
-func (resp *Response) decodeHeader(d *msgpack.Decoder) (err error) {
+func decodeHeader(d *msgpack.Decoder, buf *smallBuf) (Header, iproto.Type, error) {
 	var l int
-	d.Reset(&resp.buf)
+	var code int
+	var err error
+	d.Reset(buf)
 	if l, err = d.DecodeMapLen(); err != nil {
-		return
+		return Header{}, 0, err
 	}
+	decodedHeader := Header{}
 	for ; l > 0; l-- {
 		var cd int
-		if cd, err = resp.smallInt(d); err != nil {
-			return
+		if cd, err = smallInt(d, buf); err != nil {
+			return Header{}, 0, err
 		}
 		switch iproto.Key(cd) {
 		case iproto.IPROTO_SYNC:
 			var rid uint64
 			if rid, err = d.DecodeUint64(); err != nil {
-				return
+				return Header{}, 0, err
 			}
-			resp.RequestId = uint32(rid)
+			decodedHeader.RequestId = uint32(rid)
 		case iproto.IPROTO_REQUEST_TYPE:
-			var rcode uint64
-			if rcode, err = d.DecodeUint64(); err != nil {
-				return
+			if code, err = d.DecodeInt(); err != nil {
+				return Header{}, 0, err
 			}
-			resp.Code = uint32(rcode)
+			decodedHeader.Error = iproto.Error(code &^
+				int(iproto.IPROTO_TYPE_ERROR))
```

Comment:
Let's make this code obvious for a reader.
```suggestion
			if code & int(iproto.IPROTO_TYPE_ERROR) {
				decodedHeader.Error = iproto.Error(code &^ int(iproto.IPROTO_TYPE_ERROR))
			} else {
				decodedHeader.Error = ErrorNo
			}
```

Reply:
We need to check that IPROTO_TYPE_ERROR bit is set:
code & IPROTO_TYPE_ERROR != 0

And unset the bit:
code & (^IPROTO_TYPE_ERROR)

Or I miss something?

Reply:
Yeah, you are right.